### PR TITLE
feat: add time-boxed sessions

### PR DESCRIPTION
**Pull request by hf**
Adds time-boxed sessions. These sessions time-out after a fixed amount of time configured via the `GOTRUE_SESSIONS_TIMEBOX` property (if set).

### CODE REVIEW

File: internal/api/api.go [hunk 1]
```diff
@@ -31,6 +31,16 @@ type API struct {
...
+
+	overrideTime func() time.Time
+}
+
+func (a *API) Now() time.Time {
+	if a.overrideTime != nil {
+		return a.overrideTime()
+	}
+
+	return time.Now()
```
Comment (kangmingtay) [inline internal/api/api.go:43]: not too sure why this change is necessary, when will we need to override the time function here?

File: internal/conf/configuration.go [hunk 2]
```diff
@@ -111,6 +111,10 @@ func (a *APIConfiguration) Validate() error {
 	return nil
 }
 
+type SessionsConfiguration struct {
+	Timebox time.Duration `json:"timebox"`
```
Comment (kangmingtay) [inline internal/conf/configuration.go:115]: personal preference to make this a pointer so it defaults to `nil` if not set instead of time.Duration(0)

File: internal/api/token_refresh.go [hunk 3]
```diff
@@ -69,7 +69,15 @@ func (a *API) RefreshTokenGrant(ctx context.Context, w http.ResponseWriter, r *h
 				notAfter = *session.NotAfter
 			}
 
-			if !notAfter.IsZero() && time.Now().UTC().After(notAfter) {
+			if config.Sessions.Timebox != time.Duration(0) {
+				sessionEndsAt := session.CreatedAt.Add(config.Sessions.Timebox.Abs())
+
+				if notAfter.IsZero() || sessionEndsAt.Before(notAfter) {
```
[the remark this answers is not in the document]
Comment (kangmingtay) [reply internal/api/token_refresh.go:75]: also maybe worth adding a comment that the timebox config overwrites `session.NotAfter`

File: internal/api/api.go (hunk 1 above)
Comment (hf) [reply internal/api/api.go:43]: For tests

File: internal/conf/configuration.go (hunk 2 above)
Comment (J0) [reply internal/conf/configuration.go:115]: Should we forbid negative durations as well?

File: internal/api/token_test.go [hunk 4]
```diff
@@ -60,6 +60,40 @@ func (ts *TokenTestSuite) SetupTest() {
...
+	ts.API.handler.ServeHTTP(w, req)
+	assert.Equal(ts.T(), http.StatusBadRequest, w.Code)
+
+	var firstResult struct {
+		Error            string `json:"error"`
+		ErrorDescription string `json:"error_description"`
+	}
+
+	assert.NoError(ts.T(), json.NewDecoder(w.Result().Body).Decode(&firstResult))
+	assert.Equal(ts.T(), firstResult.Error, "invalid_grant")
```
Comment (J0) [inline internal/api/token_test.go:93]: /nit Think the [expected argument comes first](https://pkg.go.dev/github.com/stretchr/testify/assert#Equal)
```suggestion
	assert.Equal(ts.T(), "invalid_grant", firstResult.Error)
```

**Review by J0 — APPROVED** [review]

File: internal/api/api.go (hunk 1 above)
Comment (kangmingtay) [reply internal/api/api.go:43]: or even set the timebox value to 0 if it's possible in the tests

Comment (kangmingtay) [reply internal/api/api.go:43]: imo the purpose of overriding `time.Now()` is to test the behaviour that the timebox works and it revokes an expired session which is equivalent to setting the timebox to 0 such that any session expires immediately

Comment (hf) [reply internal/api/api.go:43]: There aren't too many options when testing time. Usually the best way is to provide some way to inject the clock you want to test with... Go is pretty basic so you have to use some `overrideX` style variable to control time. In other languages DI solves this pretty neatly. Other languages also allow you to modify the wall clock, but that's not really an option here as well.

I'm in favor of 0 duration as the 0 value is disallowed from the config (see the new Validate method) as it doesn't convey that much meaning -- is it disabled or not? Milliseconds are also not an option because that can cause flaky tests.

In general all tests involving time should switch to using `api.Now()` instead of `time.Now()` so we can test that behavior properly. Probably a task for the future (though once I'm done with inactivity timeout I'll rewrite all `time.Now()` in this file to use `api.Now()` instead).

**Review by kangmingtay — APPROVED** [review]
